### DISCUSSION TITLE
Submit signed challenge fails with

### DISCUSSION THREAD

**Discussion by fue-medatixx** [Digitaler-Impfnachweis/certification-apis#46] [Q&A] [answered]
Since the RU was changed yesterday, the submit request for the signed challenge has failed. Has anything been changed in the verification of the signed challenge? Testing against the RU is quite frustrating.

HTTP/1.1 401 Unauthorized
Cache-Control: no-store, must-revalidate, max-age=0
Content-Language: en
Content-Length: 2421
Content-Security-Policy: frame-src 'self'; frame-ancestors 'self'; object-src 'none';
Content-Type: text/html;charset=utf-8
Date: Sat, 29 May 2021 15:22:27 GMT
Referrer-Policy: no-referrer
Server: istio-envoy
Strict-Transport-Security: max-age=15552000; includeSubDomains; preload
X-Content-Type-Options: nosniff
X-Envoy-Decorator-Operation: keycloak-certificato-headless.ubirch-certify-demo.svc.cluster.local:80/*
X-Envoy-Upstream-Service-Time: 9
X-Frame-Options: SAMEORIGIN
X-Robots-Tag: none
X-Xss-Protection: 1; mode=block

    <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
            "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
    <html xmlns="http://www.w3.org/1999/xhtml" class="login-pf">

    <head>
        <meta charset="utf-8">
        <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/>
        <meta name="robots" content="noindex, nofollow">

                <meta name="viewport" content="width=device-width,initial-scale=1"/>
        <title>Log in to Ubirch Certify App Login</title>
        <link rel="icon" href="/auth/resources/mrpuj/login/Ubirch-simple/img/favicon.ico"/>
                <link href="/auth/resources/mrpuj/common/keycloak/node_modules/patternfly/dist/css/patternfly.min.css" rel="stylesheet"/>
                <link href="/auth/resources/mrpuj/common/keycloak/node_modules/patternfly/dist/css/patternfly-additions.min.css" rel="stylesheet"/>
                <link href="/auth/resources/mrpuj/common/keycloak/lib/zocial/zocial.css" rel="stylesheet"/>
                <link href="/auth/resources/mrpuj/login/Ubirch-simple/css/login-28042021.css" rel="stylesheet"/>
    <SCRIPT> if (typeof history.replaceState === 'function') {  history.replaceState({}, "some title", "https://id.certify.demo.ubirch.com/auth/realms/ubirch-certify/login-actions/authenticate?execution=3c2867cc-c85a-4c22-9ffa-81fdade35da6&client_id=user-access-ti&tab_id=ai3J-eyeAsc"); }</SCRIPT></head>

    <body class="">

    <div class="header">
        <a class="logo" href="https://ubirch.de">
            <img src="/auth/resources/mrpuj/login/Ubirch-simple/img/Ubirch_Logo_weiss.svg" alt="logo"></a>
        <div id="kc-header-wrapper"
             class="">Ubirch Certify App Login</div>
    </div>

    <div class="login-pf-page">
        <div id="kc-header" class="login-pf-page-header">
            <div id="kc-header-wrapper"
                 class="">Ubirch Certify App Login</div>
        </div>
        <div class="card-pf ">
            <header class="login-pf-header">
                        <h1 id="kc-page-title">        We are sorry...
</h1>
            </header>
            <div id="kc-content">
                <div id="kc-content-wrapper">

        <div id="kc-error-message">
            <p class="instruction">Certificate & Signature Verification failed</p>
        </div>

                </div>
            </div>

        </div>
    </div>
    </body>
    </html>

**Comment by mge-medisoftware** [comment]
We can confirm this: 401 Unauthorized when submitting signed challenge and public certificate. 
We use the secunet-connector for the signature-process, with public certificate C.AUT (it's not mentioned which certificate  is used in ExternalAuthenticate(), but same errorcode occurs with certificate C.SIG)

**Comment by mge-medisoftware** [comment, marked as answer]
This problem has been fixed. Now sticking with known problem using Issuer Service API in TI RU (BSNR whitelisting in TIR RU???):
"Sorry, the credential is invalid:Id not whitelisted -> Id:011111100"

**Reply by oliver-steinbrecher** [reply]
RU issues have been solved and successfully tested.

**Reply by mge-medisoftware** [reply]
yes, with some real-life BSNR it worked, thanks

**Reply by msteffenDev** [reply]
How did you solve the problem 401 Unauthorized, I tried all SMC-B certificates in ExternalAuthenticate?
Thanks

**Reply by mge-medisoftware** [reply]
The problem has been fixed serverside somehow. We encode the XAuth-challenge (string including hyphens!) base64 and execute ExternalAuthenticate() with signature type PKCS1. Then proceeding with signed challenge and public certificate C.AUT. Works on TI RU.

**Reply by twwolff** [reply]
Here (https://github.com/Digitaler-Impfnachweis/certification-apis/blob/master/SMCB-Authentication.md) is mentionned that the challenge must be hashed with sha256 and after that signed with externalAuthenticate, when do you Base64 encode the challenge ? The signatur type must be set in the optionalInput ?  when i set PKCS1 as Signatur type i receive an Error : Ungültiger Signaturtyp oder Signaturvariante. Where is the BSNR set ? i didn't see something like that in the documentation. Thank for the help.

**Reply by ManuelB** [reply]
@twwolff here is a working implementation of the whole flow:

https://github.com/ere-health/dgc-ps-app/blob/video/covid-19-ru-demo/src/main/java/health/ere/ps/service/idp/client/IdpClient.java

The part that is interesting for you is here:
https://github.com/ere-health/dgc-ps-app/blob/12c68945453b9615b58624426811c1ccd8c0c951/src/main/java/health/ere/ps/service/idp/crypto/jose4j/JsonWebSignatureWithExternalAuthentification.java#L74

Video is here:
https://www.youtube.com/watch?v=C0KfzxP-HzU

This code is GPL so you are allowed to see how it works but you are not allowed to copy and paste it.

**Reply by twwolff** [reply]
Thank for the answer, i have check the code, my code use some other tools but do the same at then end, i do not understand why i cannot submit the signed challenge and the certificate , the server is always closing the connection. any idea ? I have a test connector, is that necessary to whitelist the smcb card or something like that ? Thank for your help .

**Reply by ManuelB** [reply]
@twwolff the SMC-B has to be in the verified state. This can be done by using the CardService Verify PIN function or when you are using a Secunet Connector by clicking in the backend on the verify SMC-B pin link.

**Reply by twwolff** [reply]
Thank yes this is in verified state. (we have a secunet Test-connector).. I really don't know why it is not working... Maybe my connector is the problem.. SIS Vpn State is not connected (is that normal ?). Sorry to bother you.. Thank ageain !

**Reply by ManuelB** [reply]
@twwolff unfortunately I can just refer to the following video: https://www.youtube.com/watch?v=C0KfzxP-HzU In this video I am using a secunet PTV4 connector running in the RU to create a certificate. Feel free to use our solution and run the IdpClientTest test case. This test case will show extensive logging. If you need commercial support also feel free to contact me.

**Reply by twwolff** [reply]
reboot from connector was the solution !!! Thank

**Reply by ManuelB** [reply]
@twwolff when new routes are added to the connector a restart is required. Normally this is shown in the UI of the connector.

**Comment by msteffenDev** [comment]
We still have the problem mentioned above:

...

* Connection state changed (MAX_CONCURRENT_STREAMS updated)!
< HTTP/2 401 
< cache-control: no-store, must-revalidate, max-age=0
< content-language: en
< content-security-policy: frame-src 'self'; frame-ancestors 'self'; object-src 'none';
< content-type: text/html;charset=utf-8
< date: Fri, 04 Jun 2021 13:25:34 GMT
< referrer-policy: no-referrer
< server: istio-envoy
< strict-transport-security: max-age=15552000; includeSubDomains; preload
< x-content-type-options: nosniff
< x-envoy-decorator-operation: ruproxy.ubirch-prod.svc.cluster.local:8080/*
< x-envoy-upstream-service-time: 122
< x-frame-options: SAMEORIGIN
< x-robots-tag: none
< x-xss-protection: 1; mode=block
< content-length: 2418
< 
{ [2418 bytes data]
* Connection #0 to host id.ru.impfnachweis.info left intact
HTTP/2 401 
cache-control: no-store, must-revalidate, max-age=0
content-language: en
content-security-policy: frame-src 'self'; frame-ancestors 'self'; object-src 'none';
content-type: text/html;charset=utf-8
date: Fri, 04 Jun 2021 13:25:34 GMT
referrer-policy: no-referrer
server: istio-envoy
strict-transport-security: max-age=15552000; includeSubDomains; preload
x-content-type-options: nosniff
x-envoy-decorator-operation: ruproxy.ubirch-prod.svc.cluster.local:8080/*
x-envoy-upstream-service-time: 122
x-frame-options: SAMEORIGIN
x-robots-tag: none
x-xss-protection: 1; mode=block
content-length: 2418

    <!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional//EN\"
            \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\">
    <html xmlns=\"http://www.w3.org/1999/xhtml\" class=\"login-pf\">

    <head>
        <meta charset=\"utf-8\">
        <meta http-equiv=\"Content-Type\" content=\"text/html; charset=UTF-8\"/>
        <meta name=\"robots\" content=\"noindex, nofollow\">

                <meta name=\"viewport\" content=\"width=device-width,initial-scale=1\"/>
        <title>Log in to Ubirch Certify App Login</title>
        <link rel=\"icon\" href=\"/auth/resources/mrpuj/login/Ubirch-simple/img/favicon.ico\"/>
                <link href=\"/auth/resources/mrpuj/common/keycloak/node_modules/patternfly/dist/css/patternfly.min.css\" rel=\"stylesheet\"/>
                <link href=\"/auth/resources/mrpuj/common/keycloak/node_modules/patternfly/dist/css/patternfly-additions.min.css\" rel=\"stylesheet\"/>
                <link href=\"/auth/resources/mrpuj/common/keycloak/lib/zocial/zocial.css\" rel=\"stylesheet\"/>
                <link href=\"/auth/resources/mrpuj/login/Ubirch-simple/css/login-28042021.css\" rel=\"stylesheet\"/>
    <SCRIPT> if (typeof history.replaceState === 'function') {  history.replaceState({}, \"some title\", \"https://id.ru.impfnachweis.info/auth/realms/ubirch-certify/login-actions/authenticate?execution=fac97347-d739-43a5-adeb-1f600009e0ba&client_id=user-access-ti&tab_id=x5RQ_2Xr1Z4\"); }</SCRIPT></head>

    <body class=\"\">

    <div class=\"header\">
        <a class=\"logo\" href=\"https://ubirch.de\">
            <img src=\"/auth/resources/mrpuj/login/Ubirch-simple/img/Ubirch_Logo_weiss.svg\" alt=\"logo\"></a>
        <div id=\"kc-header-wrapper\"
             class=\"\">Ubirch Certify App Login</div>
    </div>

    <div class=\"login-pf-page\">
        <div id=\"kc-header\" class=\"login-pf-page-header\">
            <div id=\"kc-header-wrapper\"
                 class=\"\">Ubirch Certify App Login</div>
        </div>
        <div class=\"card-pf \">
            <header class=\"login-pf-header\">
                        <h1 id=\"kc-page-title\">        We are sorry...
</h1>
            </header>
            <div id=\"kc-content\">
                <div id=\"kc-content-wrapper\">


        <div id=\"kc-error-message\">
            <p class=\"instruction\">Certificate & Signature Verification failed</p>
        </div>


                </div>
            </div>

        </div>
    </div>
    </body>
    </html>

**Comment by ManuelB** [comment]
Ich hab mal unsere Lösung hier detailliert erklärt: https://youtu.be/oLrjF-OISO0 https://github.com/ere-health/dgc-ps-app
